lib/ansible/plugins/action/fetch.py

```diff
@@ -161,29 +170,63 @@ def run(self, tmp=None, task_vars=None):
 
             dest = os.path.normpath(dest)
 
+            re_encrypt_needed = False
+            local_encrypted = False
+            b_file_path = to_bytes(dest, errors='surrogate_or_strict')
+            if self._loader.path_exists(b_file_path) and self._loader.is_file(b_file_path):
+                with open(to_bytes(dest), 'rb') as f:
+                    if self._loader._vault.is_encrypted(f.read(len(b_HEADER))):
+                        f.seek(0)
+                        data = f.read()
+                        if not self._loader._vault.secrets:
+                            raise AnsibleParserError("A vault password or secret must be specified to decrypt %s" % to_native(dest))
+
+                        data, vault_id_used, vault_secret_used = self._loader._vault.decrypt_and_get_vault_id(data, filename=dest)
+                        local_encrypted = True
+                        local_checksum = checksum_s(data)
+                        re_encrypt_needed = (vault_id_used != vault_id or vault_secret_used != vault_secret)
+
             # calculate checksum for the local file
-            local_checksum = checksum(dest)
+            if not local_encrypted:
+                local_checksum = checksum(dest)
 
-            if remote_checksum != local_checksum:
+            if remote_checksum != local_checksum or local_encrypted != encrypt or re_encrypt_needed:
```

Comment:
I don't think `local_encrypted != encrypt or re_encrypt_needed` is quite right. Here's what's happening more explicitly:

```
if (
    (local_encrypted is False and encrypt is True)  # needs encrypt
    or (local_encrypted is True and encrypt is False)  # what exactly should this do? since encrypt is false, I'd expect backwards-compatible behavior and no encryption/decryption
    or re_encrypt_needed  # needs decrypt + encrypt
)
```

You could probably consolidate `re_encrypt_needed` and `local_encrypted` into a single variable, by setting `re_encrypt_needed` to None initially, and then use something more like `if encrypt or re_encrypt_needed`.

Reply:
>  or (local_encrypted is True and encrypt is False)  # what exactly should this do? since encrypt is false, I'd expect backwards-compatible behavior and no encryption/decryption

I think the most intuitive is to decrypt the file. If you had encrypt=true and change the playbook to encrypt=false, it means you no longer want the file encrypted, right?

Reply:
:thinking: maybe `encrypt` should default to `None` and decrypt the source when set to `False`. I think there should be a way to reinstate the original behavior, the feature should be additive.

Reply:
Ah, I think I finally got your point :-). Sorry it took so long :-). I really didn't consider a case when somebody would be transferring vault-encrypted file, you are right that in that case there is currently no way to preserve current behavior. It is either double encrypted, or decrypted.
Your suggestion with tri-state flag is interesting, thanks. I will think about it...

Reply:
Sorry for my contrived examples :) generally looks really good to me, just wanted to hash out some potential corner cases.

Reply:
See my comment above about the options ... but also, what happens when vault secret used is different than the one on the existing file?

Reply:
Both vault_secret and vault_id are checked, if they differ, the file is re-encrypted with the new vaules.

Reply:
but what if that is not what author wanted? we have explicit 'rekey' actions in vault for that reason.

Reply:
> Sorry for my contrived examples :) generally looks really good to me, just wanted to hash out some potential corner cases.

Don't be sorry, I am glad you pointed this out, I really didn't think about such a case and it makes sense to handle as much cases as possible correctly :-). All the comments here are extremely helpful, thanks for helping me to get this polished out.

Reply:
> but what if that is not what author wanted? we have explicit 'rekey' actions in vault for that reason.

Again, can you, please, expand on this a bit? Example use-case?
The playbook specifies secret/vault to use for encryption, so the file is encrypted with that secret/vault. What else would you expect?

Reply:
Just looking at possible scenarios and how much 'safety' should be added:
- file is encrypted with vaultid 23
- playbook prompts user for vault id/secret
- user uses/mistypes diff vaultid/secret but did not mean to rekey file
- now you have encrypted file that will fail with expected access

Reply:
@bcoca Encrypted sources are not supported, it doesn't matter whether it's the same vault ID or not, if src is encrypted and either the vault ID doesn't match or the checksums don't match, you'd get:
```
fatal: [localhost]: FAILED! => {
    "msg": "input is already encrypted"
}
```
If the vault ID doesn't match then `re_encrypt_needed` is True - but this variable name was misleading to me, it doesn't actually let you re-encrypt an already-encrypted source with a new vault ID.

https://github.com/ansible/ansible/blob/devel/lib/ansible/parsing/vault/__init__.py#L608-L609

Also similarly, the vault decrypt methods can't be called with plaintext. There are no explicit guardrails against that either, but I couldn't get this branch to accidentally decrypt plaintext.

Reply:
i was thinking destinations .. but encrypted sources also opens the issue of 'should i encrypt the encrypted'